pac4j-saml/src/main/java/org/pac4j/saml/sso/artifact/SAML2ArtifactBindingDecoder.java

```diff
@@ -0,0 +1,96 @@
+import org.opensaml.saml.saml2.metadata.ArtifactResolutionService;
+import org.opensaml.saml.saml2.metadata.IDPSSODescriptor;
+import org.opensaml.security.SecurityException;
+import org.opensaml.soap.client.http.PipelineFactoryHttpSOAPClient;
+import org.opensaml.soap.common.SOAPException;
+import org.pac4j.core.context.J2EContext;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.saml.context.SAML2MessageContext;
+import org.pac4j.saml.metadata.SAML2MetadataResolver;
+import org.pac4j.saml.transport.AbstractPac4jDecoder;
+
+/**
+ * Decodes a SAML artifact binding request by fetching the actual artifact via
+ * SOAP.
+ * 
+ * @since 3.8.0
+ */
+public class SAML2ArtifactBindingDecoder extends AbstractPac4jDecoder {
+    private final SAML2MetadataResolver idpMetadataResolver;
+
+    private final SAML2MetadataResolver spMetadataResolver;
+
+    private final SOAPPipelineProvider soapPipelineProvider;
+
+    public SAML2ArtifactBindingDecoder(final WebContext context, final SAML2MetadataResolver idpMetadataResolver,
+            final SAML2MetadataResolver spMetadataResolver, final SOAPPipelineProvider soapPipelineProvider) {
+        super(context);
+        this.idpMetadataResolver = idpMetadataResolver;
+        this.spMetadataResolver = spMetadataResolver;
+        this.soapPipelineProvider = soapPipelineProvider;
+    }
+
+    @Override
+    public String getBindingURI(SAML2MessageContext messageContext) {
+        return SAMLConstants.SAML2_ARTIFACT_BINDING_URI;
+    }
+
+    @Override
+    protected void doDecode() throws MessageDecodingException {
+        try {
+            DefaultEndpointResolver<ArtifactResolutionService> endpointResolver = new DefaultEndpointResolver<>();
+            endpointResolver.initialize();
+
+            PredicateRoleDescriptorResolver roleResolver = new PredicateRoleDescriptorResolver(
+                    idpMetadataResolver.resolve());
+            roleResolver.initialize();
+
+            SAML2MessageContext messageContext = new SAML2MessageContext();
+
+            PipelineFactoryHttpSOAPClient<SAMLObject, SAMLObject> soapClient = new PipelineFactoryHttpSOAPClient<SAMLObject, SAMLObject>() {
+                @SuppressWarnings("rawtypes")
+                @Override
+                public void send(String endpoint, InOutOperationContext operationContext)
+                        throws SOAPException, SecurityException {
+                    super.send(endpoint, operationContext);
+                    transferContext(operationContext, messageContext);
+                }
+            };
+            soapClient.setPipelineFactory(soapPipelineProvider.getPipelineFactory());
+            soapClient.setHttpClient(soapPipelineProvider.getHttpClientBuilder().buildClient());
+
+            HTTPArtifactDecoder artifactDecoder = new HTTPArtifactDecoder();
+            artifactDecoder.setHttpServletRequest(((J2EContext) context).getRequest());
```

Comment:
This is a serious issue: this won't work in Play, Vert.x and many other frameworks. Can we stay at a more abstract level (web context)?

Reply:
Unfortunately the `HTTPArtifactDecoder` requires the servlet request. It is only used to read parameters (`RelayState` and `SAMLart`), but the class is written in such a way that the request is mandatory. There is no way to modify the class (ie by overriding certain methods) to make it work without the request. It should be possible to implement most of the `HttpServletRequest` interface by wrapping the `WebContext`. I do want to use `HTTPArtifactDecoder`, because it performs most of the complicated stuff.

Reply:
Sure, I understand, but we can't put that in pac4j, this would break the pac4j core promise.

I took a look at the `HTTPArtifactDecoder`, it seems that the request is only used to get the `SAMLart` parameter. Apart from the fact that the code does not allow easy overriding, it does not seem to be used more than that.
I know it's painful, but can you copy/paste this class somehow and use the web context?

Reply:
In that case, I'd rather go for wrapping the web context in a dummy `HttpServletRequest`. IMHO `HTTPArtifactDecoder` contains way too much code to simply copy paste it. `WebContext` has enough information to implement at least the basics of `HttpServletRequest`. As you said, only `getParameter(String)` is used, but many of the other methods map just fine to `WebContext`. The ones missing can throw `UnsupportedOperationException`.

This way we can continue to use `HTTPArtifactDecoder` without having to copy paste the code (and remember to do that again and again on every upgrade of OpenSAML) and keep the contract of pac4j core.

Reply:
I understand your concerns, but how can you do wrap the web context in a `HttpServletRequest` for a framework which has none?

Reply:
`HttpServletRequest` is just an interface which you can implement. The only problem is that it has a gazillion methods, many of which will not be supported by `WebContext`, but that's fine for this use case. I actually started that way when I wrote this code, but thought it would be fine to do it like this. I wasn't aware that pac4j promises not to depend on a servlet container like this. I can implement this tomorrow.

Reply:
I've implemented an `HttpServletRequest` and `HttpSession` on top of `WebContext`. Many of the methods do not map, but the most important ones do. At least it is more than enough for `HTTPArtifactDecoder`, which actually uses `getContentType()` as well:
```java
log.debug("Beginning to decode message from HttpServletRequest");
log.debug("HttpServletRequest indicated Content-Type: {}",
        getHttpServletRequest().getContentType());
```

Reply:
OK. But I still don't see how it can work in a non-JEE environment with an import on `HttpServletRequest`? It will return a `ClassNotFoundException`...

Reply:
Well, you do need the servlet-api classes on the classpath. You only need the classes, you don't need the whole container or implementation.

Reply:
I understand, but this won't be acceptable. Can you revert your latest change? I'll merge it as is.

It will work for J2E only at first, I'll change the `HTTPArtifactDecoder` implementation later on.

Reply:
I've removed the last commit from the PR

Reply:
OK. Thanks